React/Views/SafeAreaView/RCTSafeAreaView.m

```diff
@@ -101,6 +101,17 @@ - (void)layoutSubviews
 
 - (void)setSafeAreaInsets:(UIEdgeInsets)safeAreaInsets
 {
+
+// Use the app's key window to determine Safe area insets.
+// fixes: https://github.com/facebook/react-native/issues/18177
+#if defined(__IPHONE_OS_VERSION_MAX_ALLOWED) && __IPHONE_OS_VERSION_MAX_ALLOWED >= 110000 /* __IPHONE_11_0 */
+  if (self.isSupportedByOS) {
+    if (@available(iOS 11.0, *)) {
+      safeAreaInsets = UIApplication.sharedApplication.keyWindow.safeAreaInsets;
```

Comment:
@ericlewis 
I may have some worries about this, because we'd better let system to calculate and tell us the insets, and I may think the issue is caused by `Modal`. 🤔

Reply:
@ericlewis I think this is reasonable, though I'm curious if what @zhongwuzw says is true / if there's a way of fixing this from Modal. If not, and regardless, can we move this logic from `setSafeAreaInsets` to `safeAreaInsetsIfSupportedAndEnabled` or `invalidateSafeAreaInsets `? Presumably `setSafeAreaInsets` is being called from `invalidateSafeAreaInsets`, so it seems weird to hijack the value in `setSafeAreaInsets` instead of just setting the correct value from `invalidateSafeAreaInsets`.

If it's not being called from `invalidateSafeAreaInsets`, where is it being called from, and could we fix it in that call site?

Reply:
@JoshuaGross we could move it, but after messing with this more I noticed that when nesting SafeAreaViews that it will continually add the insets from window resulting in too much padding where it shouldn’t be! I might take a different approach here... maybe look at modal itself.